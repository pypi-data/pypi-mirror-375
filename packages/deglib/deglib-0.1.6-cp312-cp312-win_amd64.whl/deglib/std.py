from deglib_cpp import Mt19937 as Mt19937cpp


class Mt19937:
    def __init__(self, seed: int = 5489):
        """
        Create a mersenne twister engine for pseudo random number generation.
        """
        self.mt19937_cpp = Mt19937cpp(seed)

    def to_cpp(self) -> Mt19937cpp:
        return self.mt19937_cpp
