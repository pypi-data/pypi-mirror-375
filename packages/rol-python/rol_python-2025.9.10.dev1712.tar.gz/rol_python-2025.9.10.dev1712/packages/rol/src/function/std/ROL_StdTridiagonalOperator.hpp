// @HEADER
// *****************************************************************************
//               Rapid Optimization Library (ROL) Package
//
// Copyright 2014 NTESS and the ROL contributors.
// SPDX-License-Identifier: BSD-3-Clause
// *****************************************************************************
// @HEADER

#ifndef ROL_STDTRIDIAGONALOPERATOR_H
#define ROL_STDTRIDIAGONALOPERATOR_H

#include "ROL_StdLinearOperator.hpp"

/** @ingroup func_group
    \class ROL::StdTridiagonalOperator
    \brief Provides the std::vector implementation to apply a linear operator,
      which encapsulates a tridiagonal matrix

    \f[ T = \begin{matrix} 
    a_0 & b_0 \\
    c_0 & a_1 & b_1 \\
        & c_1 & \ddots & \ddots \\
        &     & \ddots & a_{n-2} & b_{n-2} \\
        &     &        & c_{n-2} & a_{n-1}
    \end{pmatrix} \f]
    ---
*/


namespace ROL {

template <class Real>
class StdTridiagonalOperator : public StdLinearOperator<Real> {
 
  template <typename T> using vector = std::vector<T>;
  
private:

  const ROL::Ptr<const vector<Real> > a_; // Diagonal
  const ROL::Ptr<const vector<Real> > b_; // Superdiagonal
  const ROL::Ptr<const vector<Real> > c_; // Subdiagonal

  mutable vector<Real> dl_;
  mutable vector<Real> d_;
  mutable vector<Real> du_;
  mutable vector<Real> du2_;

  mutable vector<int>  ipiv_;

  int N_;
  
  ROL::LAPACK<int,Real>  lapack_;

  void copy(void) const {
    for(int i=0;i<N_-1;++i) {
      dl_[i]  = (*c_)[i];
      d_[i]   = (*a_)[i];
      du_[i]  = (*b_)[i]; 
    }  
    d_[N_-1] = (*a_)[N_-1];
    du2_.assign(N_-2,0.0);
  }

public:

  StdTridiagonalOperator( const ROL::Ptr<const vector<Real> > &a,
                          const ROL::Ptr<const vector<Real> > &b,
                          const ROL::Ptr<const vector<Real> > &c ) : 
    StdLinearOperator<Real>(), a_(a), b_(b), c_(c) { 

    N_ = a_->size();
    
    dl_.resize(N_-1);
    d_.resize(N_);
    du_.resize(N_-1);
    du2_.resize(N_-2);
    ipiv_.resize(N_);   
  }

  StdTridiagonalOperator( const ROL::Ptr<const vector<Real> > &a,
                          const ROL::Ptr<const vector<Real> > &b ) {
    StrdTridiagonalOperator(a,b,b);
  }
   

  virtual ~StdTridiagonalOperator() {}
 
  using StdLinearOperator<Real>::apply;
  using StdLinearOperator<Real>::applyInverse;
  using StdLinearOperator<Real>::applyAdjoint;
  using StdLinearOperator<Real>::applyAdjointInverse;

 
  virtual void apply( vector<Real> &Hv, const vector<Real> &v, Real &tol ) const {
    Hv[0] = (*a_)[0]*v[0] + (*b_)[0]*v[1];

    for( int i=1; i<N_-1; ++i ) {
      Hv[i] = (*c_)[i-1]*v[i-1] + (*a_)[i]*v[i] + (*b_)[i]*v[i+1];
    }
          
    Hv[N_-1] = (*a_)[N_-1]*v[N_-1] + (*c_)[N_-2]*v[N_-2];

  }

  virtual void applyAdjoint( vector<Real> &Hv, const vector<Real> &v, Real &tol ) const {
    Hv[0] = (*a_)[0]*v[0] + (*c_)[0]*v[1];

    for( int i=1; i<N_-1; ++i ) {
      Hv[i] = (*b_)[i-1]*v[i-1] + (*a_)[i]*v[i] + (*c_)[i]*v[i+1];
    }
          
    Hv[N_-1] = (*a_)[N_-1]*v[N_-1] + (*b_)[N_-2]*v[N_-2];
  }

  virtual void applyInverse( vector<Real> &Hv, const vector<Real> &v, Real &tol ) const {
    Hv = v;
    copy();
    int INFO;
    lapack_.GTTRF(N_,&dl_[0],&d_[0],&du_[0],&du2_[0],&ipiv_[0],&INFO); 
    lapack_.GTTRS('N',N_,1,&dl_[0],&d_[0],&du_[0],&du2_[0],&ipiv_[0],&Hv[0],N_,&INFO);
}

  virtual void applyAdjointInverse( vector<Real> &Hv, const vector<Real> &v, Real &tol ) const {
    Hv = v;
    copy();
    int INFO;
    lapack_.GTTRF(N_,&dl_[0],&d_[0],&du_[0],&du2_[0],&ipiv_[0],&INFO); 
    lapack_.GTTRS('T',N_,1,&dl_[0],&d_[0],&du_[0],&du2_[0],&ipiv_[0],&Hv[0],N_,&INFO);
  }

}; // class StdTridiagonalOperator

} // namespace ROL

#endif
