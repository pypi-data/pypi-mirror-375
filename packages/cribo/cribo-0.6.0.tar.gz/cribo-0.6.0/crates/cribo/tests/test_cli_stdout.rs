#![allow(clippy::disallowed_methods)] // insta macros use unwrap internally

use std::{env, process::Command};

use insta::{assert_snapshot, with_settings};

/// Helper function to get the path to a fixture file
fn get_fixture_path(relative_path: &str) -> String {
    let cwd = env::current_dir().expect("Failed to get current directory");
    let test_fixture_path = cwd.join("tests/fixtures").join(relative_path);
    test_fixture_path.to_string_lossy().to_string()
}

/// Run cribo with given arguments and return (stdout, stderr, `exit_code`)
fn run_cribo(args: &[&str]) -> (String, String, i32) {
    // Use the pre-built binary instead of cargo run for performance
    let cribo_exe = env!("CARGO_BIN_EXE_cribo");

    let output = Command::new(cribo_exe)
        .args(args)
        .env("RUST_LOG", "off")
        .env("CARGO_TERM_COLOR", "never")
        .env("NO_COLOR", "1")
        .output()
        .expect("Failed to execute command");

    let stdout = String::from_utf8_lossy(&output.stdout).to_string();
    let stderr = String::from_utf8_lossy(&output.stderr).to_string();
    let exit_code = output.status.code().unwrap_or(-1);

    (stdout, stderr, exit_code)
}

/// Filters for normalizing paths in snapshots
fn get_cli_filters() -> Vec<(&'static str, &'static str)> {
    vec![
        // Normalize file paths - Unix/macOS
        (r"/Volumes/workplace/[^\s]+", "<WORKSPACE>"),
        (r"/home/[^/]+/[^\s]+", "<WORKSPACE>"),
        (r"/Users/[^/]+/[^\s]+", "<WORKSPACE>"),
        // Normalize file paths - Windows
        (r"\\\\?[A-Z]:\\[^\s]+", "<WORKSPACE>"),
        (r"[A-Z]:\\[^\s]+", "<WORKSPACE>"),
        (r"[A-Z]:/[^\s]+", "<WORKSPACE>"),
        // Normalize cargo paths - Unix/macOS
        (r"/Users/[^/]+/\.cargo/[^\s]+", "<CARGO>"),
        (r"/home/[^/]+/\.cargo/[^\s]+", "<CARGO>"),
        // Normalize cargo paths - Windows
        (r"\\\\?C:\\Users\\[^\\]+\\\.cargo\\[^\s]+", "<CARGO>"),
        (r"C:\\Users\\[^\\]+\\\.cargo\\[^\s]+", "<CARGO>"),
        // Normalize temporary paths - Unix/macOS
        (r"/var/folders/[^/]+/[^/]+/T/[^\s]+", "<TMP>"),
        (r"/tmp/[^\s]+", "<TMP>"),
        // Normalize temporary paths - Windows
        (r"\\\\?C:\\temp\\[^\s]+", "<TMP>"),
        (r"\\\\?C:\\Windows\\Temp\\[^\s]+", "<TMP>"),
        // Normalize GitHub Actions paths
        (r"/home/runner/work/[^\s]+", "<WORKSPACE>"),
        (r"D:\\a\\[^\s]+", "<WORKSPACE>"),
        (r"C:\\hostedtoolcache\\[^\s]+", "<WORKSPACE>"),
        // Normalize content hashes that might vary across platforms
        (r"__cribo_[a-f0-9]{6,}", "__cribo_<HASH>"),
        // Normalize timestamps if any
        (r"\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}", "<TIMESTAMP>"),
        // Remove any remaining cargo output (should be minimal with --quiet)
        (r"(?m)^\s*Compiling [^\n]*\n", ""),
        (r"(?m)^\s*Finished [^\n]*\n", ""),
        (r"(?m)^\s*Blocking waiting for file lock[^\n]*\n", ""),
        (r"(?m)^\s*warning: [^\n]*unused manifest key[^\n]*\n", ""),
        // Normalize OS-specific error messages (keep structure, normalize message)
        (
            r"The system cannot find the file specified\. \(os error (\d+)\)",
            "No such file or directory (os error $1)",
        ),
        // Normalize Windows executable names
        (r"cribo\.exe", "cribo"),
        // Normalize line endings
        (r"\r\n", "\n"),
        (r"\r", "\n"),
        // Normalize module paths in bundled code
        (r"# Bundle from: [^\n]+", "# Bundle from: <MODULE_PATH>"),
    ]
}

#[test]
fn test_stdout_flag_help() {
    let (stdout, _, exit_code) = run_cribo(&["--help"]);

    // Should succeed
    assert_eq!(exit_code, 0);

    // Check help contains stdout flag
    assert!(stdout.contains("--stdout"));
    assert!(stdout.contains("Output bundled code to stdout instead of a file"));
}

#[test]
fn test_stdout_conflicts_with_output() {
    let (_, stderr, exit_code) = run_cribo(&[
        "--entry",
        "nonexistent.py",
        "--output",
        "output.py",
        "--stdout",
    ]);

    // Should fail
    assert_ne!(exit_code, 0);

    with_settings!({
        filters => get_cli_filters(),
    }, {
        assert_snapshot!("stdout_conflicts_with_output_stderr", stderr);
    });
}

#[test]
fn test_missing_output_and_stdout_flags() {
    let (_, stderr, exit_code) = run_cribo(&["--entry", "nonexistent.py"]);

    // Should fail
    assert_ne!(exit_code, 0);

    with_settings!({
        filters => get_cli_filters(),
    }, {
        assert_snapshot!("missing_output_and_stdout_stderr", stderr);
    });
}

#[test]
fn test_stdout_bundling_functionality() {
    let (stdout, stderr, exit_code) = run_cribo(&[
        "--entry",
        &get_fixture_path("simple_project/main.py"),
        "--stdout",
    ]);

    // Should succeed
    assert_eq!(exit_code, 0, "Command failed with stderr: {stderr}");

    with_settings!({
        filters => get_cli_filters(),
    }, {
        assert_snapshot!("stdout_bundling_output", stdout);
        assert_snapshot!("stdout_bundling_stderr", stderr);
    });

    // Ensure no log messages in stdout
    assert!(!stdout.contains("INFO"));
    assert!(!stdout.contains("WARN"));
    assert!(!stdout.contains("ERROR"));
}

#[test]
fn test_stdout_with_verbose_separation() {
    let (stdout, stderr, exit_code) = run_cribo(&[
        "--entry",
        &get_fixture_path("simple_project/main.py"),
        "--stdout",
        "-v",
    ]);

    // Should succeed
    assert_eq!(exit_code, 0);

    with_settings!({
        filters => get_cli_filters(),
    }, {
        assert_snapshot!("stdout_verbose_output", stdout);
        assert_snapshot!("stdout_verbose_stderr", stderr);
    });

    // Stdout should only contain Python code
    assert!(!stdout.contains("INFO"));
    assert!(!stdout.contains("Starting Cribo"));
}

#[test]
fn test_stdout_with_requirements() {
    let (stdout, stderr, exit_code) = run_cribo(&[
        "--entry",
        &get_fixture_path("simple_project/main.py"),
        "--stdout",
        "--emit-requirements",
    ]);

    // Should succeed
    assert_eq!(exit_code, 0);

    with_settings!({
        filters => get_cli_filters(),
    }, {
        assert_snapshot!("stdout_requirements_output", stdout);
        assert_snapshot!("stdout_requirements_stderr", stderr);
    });
}

#[test]
fn test_stdout_mode_preserves_bundled_structure() {
    let (stdout, _, exit_code) = run_cribo(&[
        "--entry",
        &get_fixture_path("simple_project/main.py"),
        "--stdout",
    ]);

    // Should succeed
    assert_eq!(exit_code, 0);

    // The bundled structure assertions will be in the snapshot itself
    with_settings!({
        filters => get_cli_filters(),
    }, {
        assert_snapshot!("stdout_bundled_structure", stdout);
    });
}

#[test]
fn test_stdout_error_handling() {
    let (stdout, stderr, exit_code) = run_cribo(&["--entry", "nonexistent_file.py", "--stdout"]);

    // Should fail
    assert_ne!(exit_code, 0);

    with_settings!({
        filters => get_cli_filters(),
    }, {
        assert_snapshot!("stdout_error_stdout", stdout);
        assert_snapshot!("stdout_error_stderr", stderr);
    });

    // Stdout should be empty or minimal
    assert!(stdout.is_empty() || stdout.len() < 100);
}

#[test]
fn test_directory_entry_with_main_py() {
    let (stdout, _, exit_code) = run_cribo(&[
        "--entry",
        &get_fixture_path("directory_entry_main"),
        "--stdout",
    ]);

    // Should succeed
    assert_eq!(exit_code, 0);

    with_settings!({
        filters => get_cli_filters(),
    }, {
        assert_snapshot!("directory_entry_main_stdout", stdout);
    });

    // Should contain code from __main__.py, not __init__.py
    assert!(stdout.contains("Running from __main__.py"));
    assert!(!stdout.contains("This is __init__.py"));
}

#[test]
fn test_directory_entry_with_init_py_only() {
    let (stdout, _, exit_code) = run_cribo(&[
        "--entry",
        &get_fixture_path("directory_entry_init"),
        "--stdout",
    ]);

    // Should succeed
    assert_eq!(exit_code, 0);

    with_settings!({
        filters => get_cli_filters(),
    }, {
        assert_snapshot!("directory_entry_init_stdout", stdout);
    });

    // Should contain code from __init__.py
    assert!(stdout.contains("Running from __init__.py as fallback"));
}

#[test]
fn test_directory_entry_empty_fails() {
    let (_, stderr, exit_code) = run_cribo(&[
        "--entry",
        &get_fixture_path("directory_entry_empty"),
        "--stdout",
    ]);

    // Should fail
    assert_ne!(exit_code, 0);

    with_settings!({
        filters => get_cli_filters(),
    }, {
        assert_snapshot!("directory_entry_empty_stderr", stderr);
    });

    // Should contain appropriate error message
    assert!(stderr.contains("does not contain __main__.py or __init__.py"));
}
