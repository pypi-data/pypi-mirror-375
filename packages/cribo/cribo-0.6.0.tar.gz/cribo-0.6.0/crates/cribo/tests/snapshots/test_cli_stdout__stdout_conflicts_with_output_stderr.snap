---
source: crates/cribo/tests/test_cli_stdout.rs
expression: stderr
---
error: the argument '--output <OUTPUT>' cannot be used with '--stdout'

Usage: cribo --entry <ENTRY> --output <OUTPUT>

For more information, try '--help'.
